inc/Smartling/Helpers/UserHelper.php

```
@@ -0,0 +1,55 @@
+<?php
+
+namespace Smartling\Helpers;
+
+use Smartling\DbAl\SmartlingToCMSDatabaseAccessWrapperInterface;
+
+class UserHelper {
+    use LoggerSafeTrait;
+
+    private const ADMINISTRATOR = 'administrator';
+    private const EDITOR = 'editor';
+
+    public function __construct(
+        private SmartlingToCMSDatabaseAccessWrapperInterface $db,
+        private WordpressFunctionProxyHelper $wp,
+    ) {
+    }
+
+    public function asAdministratorOrEditor(callable $function): mixed
+    {
+        $originalUser = $this->wp->wp_get_current_user();
+        if ($originalUser instanceof \WP_User && array_intersect([self::ADMINISTRATOR, self::EDITOR], $originalUser->roles)) {
+            $this->getLogger()->debug("Current user userId={$originalUser->ID} is administrator or editor, no impersonation");
+            return $function();
+        }
+
+        $privilegedId = $this->getAdministratorOrEditorId();
+        $originalUserId = $this->wp->get_current_user_id();
+
+        if ($privilegedId === null) {
+            $this->getLogger()->warning("Unable to get administrator or editor for blogId=" . $this->wp->get_current_blog_id() . ', running as original userId=' . $originalUserId);
+            return $function();
+        }
```

Maybe throw here? Sounds like for the client's code it's not expected to proceed with non privileged user in `asAdministratorOrEditor` .

---

I had my doubts here, no huge difference for now, will throw

---

I mean, if it doesn't throw then probably client will face the same issue as before, right? I don't insist on throwing, just trying think outloud.

---

Yes, it's better to throw instead of trying to continue as user without permissions, logs will be clearer